dvc/repo/diff.py

```diff
@@ -101,17 +123,25 @@ def _to_checksum(output):
     for stage in repo.stages:
         for output in stage.outs:
             if _exists(output):
-                yield _to_path(output), _to_checksum(output)
-                if output.is_dir_checksum:
-                    yield from _dir_output_paths(repo_tree, output)
+                yield_output = no_targets or output.path_info.isin_or_eq(
+                    targets
+                )
+                if yield_output:
+                    yield _to_path(output), _to_checksum(output)
+
+                if output.is_dir_checksum and (
+                    yield_output or output.path_info.isinside(targets)
+                ):
+                    yield from _dir_output_paths(repo_tree, output, targets)
```

Comment:
`isin()` will not cover every case here, I think we need to use `path_info.overlaps(...)` instead.

Let's say, `output` is a tracked directory `dir/`, which contains a file `foo`. `output.path_info` here will contain the path to the tracked dir: `dir/`.

If our target is `dir/foo`, `output.path_info.isin_or_eq(target)` will be False. In this case the output `dir/` itself should not be yielded (which you have correctly accounted for already), but the file `dir/foo` within `output` does need to be yielded (meaning we still have to call `_dir_output_paths(...)`